packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts

```diff
@@ -718,7 +718,7 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
       return null;
     }
     const emitted = emittedRef.expression;
-    if (emitted instanceof WrappedNodeExpr) {
+    if (emitted instanceof WrappedNodeExpr && emitted.node.text !== inContext.name?.text) {
```

Comment:
Instead of comparing by some string contents, you'd be better off comparing the declaration by reference equality:

```suggestion
    if (emitted instanceof WrappedNodeExpr && emitted.node !== inContext.name) {
```

This is an improvement especially because `instance WrappedNodeExpr` narrows `emitted` to `WrappedNodeExpr<any>`, so the `emitted.node` is an `any` type that may not have a `text` property, and then comparing against an optional-chained `name?.text` might get weird with undefineds in an unforeseen way. In the compiler-cli codebase there is no code size constraint, so we tend to avoid using `?.` in such direct comparisons.

Reply:
Additionally, I think I'd push this condition into a dedicated early return block:

```ts
if (emitted instanceof WrappedNodeExpr) {
  if (emitted.node === inContext.name) {
    // Suppress self-imports since components do not have to import themselves.
    return null;
  }
  // ...
}
```

as this makes the desired return value for this case much more explicit.